### PR TITLE
feat: Retry pending nodes

### PR DESCRIPTION
**Pull request by jamhed**
Checklist:

* [a] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlike to be merged.
* [ ] Optional. I've added My organization is added to the README.
* [x] I've signed the CLA and required builds are green.

### CODE REVIEW

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -1338,6 +1338,16 @@ func (woc *wfOperationCtx) executeTemplate(nodeName string, orgTmpl wfv1.Templat
 		if err != nil {
 			return woc.markNodeError(retryNodeName, err), err
 		}
+		if lastChildNode != nil && lastChildNode.Pending() && processedTmpl.GetType() == wfv1.TemplateTypeContainer {
+			_, err := woc.createWorkflowPod(lastChildNode.Name, *processedTmpl.Container, processedTmpl, false)
```
Comment (alexec) [inline workflow/controller/operator.go]: this is a significant piece of code - can you please add some tests?

File: workflow/controller/operator.go [hunk 2]
```diff
@@ -1375,6 +1389,9 @@ func (woc *wfOperationCtx) executeTemplate(nodeName string, orgTmpl wfv1.Templat
 		err = errors.Errorf(errors.CodeBadRequest, "Template '%s' missing specification", processedTmpl.Name)
 		return woc.initializeNode(nodeName, wfv1.NodeTypeSkipped, templateScope, orgTmpl, boundaryID, wfv1.NodeError, err.Error()), err
 	}
+	if apierr.IsForbidden(err) {
```
Comment (alexec) [inline workflow/controller/operator.go]: I'm assuming we're assuming that forbidden means - you don't have enough memory or similar?

File: workflow/controller/operator.go [hunk 3]
```diff
@@ -1360,7 +1370,11 @@ func (woc *wfOperationCtx) executeTemplate(nodeName string, orgTmpl wfv1.Templat
 
 	switch processedTmpl.GetType() {
 	case wfv1.TemplateTypeContainer:
-		node, err = woc.executeContainer(nodeName, templateScope, processedTmpl, orgTmpl, boundaryID)
+		if node != nil && node.Pending() {
+			_, err = woc.createWorkflowPod(node.Name, *processedTmpl.Container, processedTmpl, false)
+		} else {
```
Comment (simster7) [inline workflow/controller/operator.go]: Could we move this logic inside of `woc.executeContainer` please? Or is there a reason as to why you kept it here?

File: workflow/controller/operator.go (hunk 2 above)
Comment (jamhed) [reply workflow/controller/operator.go]: @alexec, yes, we're assuming forbidden means there are not enough resources to schedule a pod.

File: workflow/controller/operator.go (hunk 1 above)
Comment (jamhed) [reply workflow/controller/operator.go]: @alexec Yeah, I've been thinking to add e2e tests, and it requires to set a namespace with quota, let me try.

File: test/e2e/functional_test.go [hunk 4]
```diff
@@ -20,6 +21,11 @@ type FunctionalSuite struct {
 	fixtures.E2ESuite
 }
 
+func (s *FunctionalSuite) TearDownSuite() {
+	s.E2ESuite.DeleteResources(fixtures.Label)
```
Comment (alexec) [inline test/e2e/functional_test.go]: huh, should not need this, how odd

File: test/e2e/manifests/mysql.yaml [hunk 5]
```diff
@@ -267,6 +267,15 @@ subjects:
 apiVersion: v1
 data:
   config: |
+    executor:
```
Comment (simster7) [inline test/e2e/manifests/mysql.yaml]: Are these file auto-generated using `make manifest` (@alexec)? If you edited these files manually they might get overwritten next time someone runs it

Comment (jamhed) [reply test/e2e/manifests/mysql.yaml]: @simster7 yeah, I edited them manually. Let me check if I can find a source

Comment (alexec) [reply test/e2e/manifests/mysql.yaml]: @simster7 is correct. These are based on manifests/... and test/e2e/manifest/...

Comment (jamhed) [reply test/e2e/manifests/mysql.yaml]: @simster7 @alexec should be okay by now.

**Comment by simster7** [comment]
Please hold off on merging this for a bit, want to take another look

**Comment by jamhed** [comment]
@simster7 how is it going?

File: workflow/controller/operator.go [hunk 6]
```diff
@@ -1338,6 +1338,16 @@ func (woc *wfOperationCtx) executeTemplate(nodeName string, orgTmpl wfv1.Templat
 		if err != nil {
 			return woc.markNodeError(retryNodeName, err), err
 		}
+		if lastChildNode != nil && lastChildNode.Pending() && processedTmpl.GetType() == wfv1.TemplateTypeContainer {
```
Comment (simster7) [inline workflow/controller/operator.go]: Why do we only look at `wfv1.TemplateTypeContainer`? Both `wfv1.TemplateTypeScript` and `wfv1.TemplateTypeResource` also execute Pods.

File: workflow/controller/operator.go [hunk 7]
```diff
@@ -1338,6 +1338,16 @@ func (woc *wfOperationCtx) executeTemplate(nodeName string, orgTmpl wfv1.Templat
...
+			_, err := woc.createWorkflowPod(lastChildNode.Name, *processedTmpl.Container, processedTmpl, false)
+			if apierr.IsForbidden(err) {
+				return woc.markNodePending(lastChildNode.Name, err), nil
```
Comment (simster7) [inline workflow/controller/operator.go]: This return statement is incorrect. We should only return the `NodeStatus` of the parent node here (`retryParentNode`), not of the child node. The code as is would be a large bug.

File: workflow/controller/operator.go [hunk 8]
```diff
@@ -1338,6 +1338,16 @@ func (woc *wfOperationCtx) executeTemplate(nodeName string, orgTmpl wfv1.Templat
...
+			}
+			if err != nil {
+				return woc.markNodeError(lastChildNode.Name, err), err
```
Comment (simster7) [inline workflow/controller/operator.go]: Same as above.

File: workflow/controller/operator.go [hunk 9]
```diff
@@ -1338,6 +1338,16 @@ func (woc *wfOperationCtx) executeTemplate(nodeName string, orgTmpl wfv1.Templat
...
+			}
+			return woc.markNodePhase(lastChildNode.Name, wfv1.NodeRunning), nil
```
Comment (simster7) [inline workflow/controller/operator.go]: Same as above.

**Comment by jamhed** [comment]
> I would suggest refactoring the code so that an attempt to recreate the pod is considered as a full retry. 

@simster7,  these two use cases are different: to retry a failed pod, and retry the submission of the pod, so it isn't related to retryStrategy.limit, as semantic is different.

Suppose we have another parameter under retryStrategy to limit the number of resubmissions. Then imagine you have two pods, both of them require full namespace resources. Then there would be no way to tell when the first one is completed, thus making impossible to limit the number of resubmissions upfront.

Therefore, it's not a refactoring per se, but rather completely different functionality, and this is not what we want to have.

File: workflow/controller/operator.go (hunk 6 above)
Comment (jamhed) [reply workflow/controller/operator.go]: @simster7, is it possible to specify limits for these types? It's actually a requirement, as if you have limits on a namespace, it's impossible to submit a pod without limits defined. There is a workaround, thouth, LimitRange object to set default values. Also, I'd suggest if this functionality is required (to retry these types on submission failure), to make it another story.

Comment (simster7) [reply workflow/controller/operator.go]: > is it possible to specify limits for these types?

Do you mean resource limits on the k8s? Both of those template types execute pods (to k8s there is no difference between a `TemplateTypeContainer` and a `TemplateTypeScript`).

> Also, I'd suggest if this functionality is required (to retry these types on submission failure), to make it another story.

Why do you think so? From my perspective retrying Script templates is the same as retrying Container templates.

**Comment by jamhed** [comment]
@simster7, to me the difference is substantial: in one case pod exists (=consumes resources), and in another pod doesn't exist (=no resources allocated). I remember once our cluster went down because it had unlimited retries in retryStrategy, and there were an error in pod itself, so I definitely do not want to retry resubmission the same way I want to retry failed pods.

To make it perfect we need to have specific error, like QuotaError, and this requires to understand the error returned by Kubernetes API in more detail (=not how it is right now), and I sort of implemented it by passing `forbidden` to caller side, instead of wrapping it into argo error losing the details required.

To lump together submission errors (forbidden) and all other errors will make it unusable for our case (we have namespaces with limits by default for all our teams).

Please help me understand why do you want to limit the number of resubmissions due to `forbidden` reason. The only problem I see here it can overload Kubernetes API, but in that case I would prefer to rate limit resubmissions instead of limiting the number of them. However, it requires a different strategy (workflow controller needs to receive periodic updates somehow), and afaik this is not available as of now.

So, the approach suggested is not a perfect solution, but rather a practical compromise, as this issue is really a show-stopper for us (and it seems not only for us).

**Comment by jamhed** [comment]
@simster7 @alexec, so you have a slack, zoom or similar available? I really think it makes sense to discuss it live.

**Comment by jamhed** [comment]
@simster7, wrt to script tasks, here is the definition: https://github.com/argoproj/argo/blob/master/examples/retry-script.yaml
It looks custom to me (e.g. no way to set resource limits), unlike workflow template definition (=standard kubernetes pod type with resource limits available).

File: workflow/controller/operator.go [hunk 10]
```diff
@@ -1338,6 +1338,16 @@ func (woc *wfOperationCtx) executeTemplate(nodeName string, orgTmpl wfv1.Templat
 		if err != nil {
 			return woc.markNodeError(retryNodeName, err), err
 		}
+		if lastChildNode != nil && lastChildNode.Pending() && processedTmpl.GetType() == wfv1.TemplateTypeContainer {
+			_, err := woc.createWorkflowPod(lastChildNode.Name, *processedTmpl.Container, processedTmpl, false)
+			if apierr.IsForbidden(err) {
+				return woc.markNodePending(retryParentNode.Name, err), nil
```
Comment (simster7) [inline workflow/controller/operator.go]: Also, just noticed this again. The correct code here should be:

- Mark the child node with the desired phase and message
- Return the parent node

Before you would mark and return the child node and now you mark and return the parent node, both of which are not desired. Sorry if my last comment wasn't clear.

**Comment by simster7** [comment]
> @simster7 @alexec, so you have a slack, zoom or similar available? I really think it makes sense to discuss it live.

Sure! I'm `@simon` on the Argo slack

**Comment by simster7** [comment]
> @simster7, wrt to script tasks, here is the definition: https://github.com/argoproj/argo/blob/master/examples/retry-script.yaml
It looks custom to me (e.g. no way to set resource limits), unlike workflow template definition (=standard kubernetes pod type with resource limits available).

I see, sorry I had confused "ResourceQuotas" with "ResourceLimits"

File: workflow/controller/operator.go (hunk 10 above)
Comment (simster7) [reply workflow/controller/operator.go]: Something like:

```go
				_ = woc.markNodePending(lastChildNode.Name, err)
                                return retryParentNode, nil
```

**Comment by jamhed** [comment]
> Do you see this as a problem?

@simster7, I do, I'd suggest having a feature toggle for now, and make it default in future releases.

**Comment by simster7** [comment]
> @simster7, I do, I'd suggest having a feature toggle for now (I'm thinking workflow label/annotation?), and make it default in future releases (because tbh this is what I would expect from pod orchestrator). What do you think?

Let's do it then! Perhaps as a label in `retryStrategy`? Something like `retryStrategy.recreatePendingPods: true`?

Once this feature toggle is set and https://github.com/argoproj/argo/pull/2385#discussion_r391817108 is addressed we can merge this in.

File: workflow/controller/operator.go (hunk 2 above)
Comment (simster7) [reply workflow/controller/operator.go]: Can we move this inside the `if err != nil {`

File: workflow/controller/operator.go [hunk 11]
```diff
@@ -1338,6 +1351,19 @@ func (woc *wfOperationCtx) executeTemplate(nodeName string, orgTmpl wfv1.Templat
 		if err != nil {
 			return woc.markNodeError(retryNodeName, err), err
 		}
+		if lastChildNode != nil && lastChildNode.Pending() && processedTmpl.GetType() == wfv1.TemplateTypeContainer {
```
Comment (simster7) [inline workflow/controller/operator.go]: Should we check `isResubmitAllowed` here too?

**Comment by simster7** [comment]
Hey @jamhed. I just made three changes:

1. I moved the flag out of `retryStrategy` (and renamed it to `resubmitPendingPods`) and into the `template` spec. The logic for this is simple (and based on your comments): resubmitting pods is different than retrying nodes. Suppose you are a user who does not want to retry a failed pod, but wants to resubmit a pod under resource quotas. If the flag to do so is inside `retryStrategy`, you will inadvertently also be enabling retry logic, which you may not want. If you do not want retry logic, you will also be forced to add a `limit: 1` flag to avoid your container from being retried indefinitely (I can see you had to resort to that in your test). This is not obvious to the user and could result in a user accidentally allowing their container to be retried indefinitely, which is dangerous. 

1. I greatly simplified the resubmission logic. Now all of it fits inside `executeContainer` (thanks in part by moving this out of `retryStrategy`).

1. I edited your test to reflect this, and changed the E2E testing fixture to accommodate `resourcequotas` now being part of tests.

Could you please take a look at the PR now and let me know of any comments you may have? If you OK this, we'll merge it in.

**Comment by jamhed** [comment]
@simster7 it was there for a reason :) i'm quite certain it won't work this way for templates with retryStrategy.limit = 1

**Pull request merged**